### PR TITLE
XDR: fsync on close (also bucket dir), with option to disable.

### LINKED ISSUE #2202: Failure to fsync on XDR stream close
As noted in #2123 and #1395, we are seeing more instances of "missing buckets" than we expect from mere misconfiguration (which sometimes causes it, but not always).

On investigation with a node operator, we discovered a state that is best understood to be caused by our failure to fsync bucket writes:

  - Bucket went missing at moment of node crash (cloud provider killed host)
  - SQL database matched public network state (postgres WAL integrity maintained)
  - Once restored, subsequent buckets _diverge_ with difference appearing just to be missing entries or updates from lower levels (i.e. some merge restarted against a zero-sized bucket)

There are three necessary fixes for this. This bug tracks one of them: start properly fsync()'ing when we close a bucket (or any XDR file) so we don't lose them even if the OS crashes.

The other two fixes are better diagnostics (tracked with #2123) and verification of hashes while _reading_ buckets (#2203).

### PR DESCRIPTION
**Pull request by graydon**
# Description

Resolves #2202

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [x] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

Performance is worse but not hugely. Benchmarks that hammer on the bucket system non-stop run about 10x slower wall time on local NVMe and 100x worse on a spindle. And at the _very smallest_ end of the merge times, slowdown is indeed also 10x-100x: 1ms merge becomes 10ms on NVMe or even 100ms on spindle. At level 0, the 100x case is still only taking 1-3% of allowed time for the merge -- we had a ton of headroom -- but it's worse for sure.

However, it also amortizes-away pretty smoothly as we get into bigger merges. By level 4 we're back down to 0.03% of available time and at level 5 down to 0.009%. That is: the fsync is adding a definitely-noticable 100ms to a merge running on the spindle, but it's not adding _proportionally_ much more as we scale to _somewhat_ larger buckets.

To measure whether this vanishes into the noise _completely_ when scaling up (or if, say, full buffers / actually having some real data to write changes things again), I made a synthetic 1.4gb XDR test and ran it against the spindle in a loop. This gives not super wonderful but also not super deadly results. The large files _do_ definitely show different effects that are not "constant 100ms totally vanishing into the noise":

~~~
2019-07-18T14:56:41.781 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 26094ms
2019-07-18T14:56:52.930 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 11149ms
2019-07-18T14:57:28.145 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 35214ms
2019-07-18T14:57:40.637 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 12491ms
2019-07-18T14:58:01.238 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 20600ms
2019-07-18T14:58:13.760 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 12522ms
2019-07-18T14:58:44.089 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 30328ms
2019-07-18T14:58:56.132 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 12042ms
2019-07-18T14:59:24.591 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 28458ms
2019-07-18T14:59:36.892 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 12301ms
2019-07-18T14:59:56.876 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 19983ms
2019-07-18T15:00:09.571 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 12694ms
2019-07-18T15:00:41.030 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 31459ms
2019-07-18T15:00:52.665 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 11634ms
2019-07-18T15:01:13.054 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 20388ms
2019-07-18T15:01:24.972 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 11917ms
2019-07-18T15:01:54.835 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 29863ms
2019-07-18T15:02:06.456 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 11620ms
2019-07-18T15:02:26.358 <test> [Fs INFO] wrote 1410837608 bytes to fsync file in 19901ms
2019-07-18T15:02:39.802 <test> [Fs INFO] wrote 1410837608 bytes to no-fsync file in 13444ms
~~~

I.e. varies between about 1.5x and 3x.

Back in the world of NVMe (which we certainly recommend people use for their buckets) this sort of thing is much less severe, though still measurable:

~~~
2019-07-18T15:08:01.569 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11609ms
2019-07-18T15:08:15.889 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 14319ms
2019-07-18T15:08:28.031 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 12141ms
2019-07-18T15:08:39.261 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11230ms
2019-07-18T15:08:51.141 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11880ms
2019-07-18T15:09:02.121 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 10979ms
2019-07-18T15:09:14.038 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11916ms
2019-07-18T15:09:25.084 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11045ms
2019-07-18T15:09:37.012 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11927ms
2019-07-18T15:09:48.308 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11296ms
2019-07-18T15:10:00.282 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11973ms
2019-07-18T15:10:11.388 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11105ms
2019-07-18T15:10:23.253 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11864ms
2019-07-18T15:10:34.470 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11217ms
2019-07-18T15:10:46.402 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11931ms
2019-07-18T15:10:57.530 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11127ms
2019-07-18T15:11:09.532 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 12001ms
2019-07-18T15:11:20.751 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11219ms
2019-07-18T15:11:32.724 <test> [Fs INFO] wrote 1410803760 bytes to fsync file in 11972ms
2019-07-18T15:11:44.059 <test> [Fs INFO] wrote 1410803760 bytes to no-fsync file in 11335ms
~~~

### CODE REVIEW

**Comment by MonsieurNicolas** [comment]
I think that the file and dir flush code have to be moved into fs.cpp as there is non portable things in there: (at a minimum turn those into noop on platforms that don't support those things) on Windows, you have to rename with the `MOVEFILE_WRITE_THROUGH` flag for https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-movefileexa and for files, it's https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-flushfilebuffers (on a native handle)

**Comment by graydon** [comment]
Pushed update to have windows variants fleshed out. Will do some testing on OSX and BSD tomorrow.

**Comment by graydon** [comment]
OSX uncovered a bug. pushed fix. moving to BSD.

**Comment by graydon** [comment]
BSD works (well -- there's a failure but it is also on master, no change with this PR). @MonsieurNicolas  I think this is ready to land unless you have any other nits.

File: src/main/Config.h [hunk 1]
```diff
@@ -198,6 +198,17 @@ class Config : public std::enable_shared_from_this<Config>
...
+    // If set to true, writing an XDR file (a bucket or a checkpoint) will not
+    // be followed by an fsync on the file. This in turn means that XDR files
+    // (which hold the canonical state of the ledger) may be corrupted if the
+    // operating system suddenly crashes or loses power, causing the node to
+    // diverge and get stuck on restart, or potentially even publish bad
+    // history. This option only exists as an escape hatch if the local
+    // filesystem is so unusably slow that you prefer operating without
+    // durability guarantees. Do not set it to true unless you're very certain
+    // you want to make that trade.
+    bool DISABLE_XDR_FSYNC;
```
Comment (MonsieurNicolas) [inline src/main/Config.h]: forgot to add support to it in the config file (parsing code) as well as entry in example.cfg

Comment (graydon) [reply src/main/Config.h]: Fixed.

**Comment by MonsieurNicolas** [comment]
r+ 7cafbba463a881d3fef63ef82582e521b101e27b

**Pull request merged**